File: 2561_maximum_sum_of_distinct_subarrays_with_length_k.cpp
class Solution {
public:
    long long maximumSubarraySum(vector<int>& v, int w) {
        bitset<100001> b;
        long long s = 0, res = 0;
        for (int i = 0, j = 0, k = 0; j < v.size(); j++) {
            s += v[j];
            while (i < j && (k >= w || b[v[j]])) b[v[i]] = 0, s -= v[i++], k--;
            b[v[j]] = 1, k++;
            if (k == w) res = max(res, s);
        }
        return res;
    }
};

File: 2561_maximum_sum_of_distinct_subarrays_with_length_k.java
class Solution {
    public long maximumSubarraySum(int[] v, int w) {
        boolean[] mark = new boolean[100001];
        long sum = 0, maxSum = 0;
        for (int i = 0, j = 0, cnt = 0; j < v.length; j++) {
            sum += v[j];
            while (i < j && (cnt >= w || mark[v[j]])) {
                sum -= v[i];
                mark[v[i++]] = false;
                cnt--;
            }
            mark[v[j]] = true;
            cnt++;
            if (cnt == w) maxSum = Math.max(maxSum, sum);
        }
        return maxSum;
    }
}
